### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] errors not reported when using unnecessary literal assertions on `const` variable declarations

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#8566]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&code=MYewdgzgLgBAZiEMC8MA8BmAfBg3EA

### Repro Code

```TypeScript
const foo = <3>3;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{}
```

### Expected Result

`3` should be reported as an unnecessary type assertion, since removing it has no effect.

### Actual Result

The type assertion was not reported.

### Additional Info

Raised originally in #8458 and solution opened in [#8558](https://github.com/typescript-eslint/typescript-eslint/pull/8558). Breaking this out from the original issue into a separate issue.

**Comment by auvred** [comment]
Look like a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/4485

**Comment by bradzacher** [comment]
This is working as intended.

Its not always safe to remove an assertion to a literal type as there are cases where TS will widen the type - so the rule ignores assertions to literal types (and tuple types).

[example widening case](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUBKRBvAUIgxAJwFMoQikBGAblwF9dcIEBnKRAD0QF5lVMdAPRDCiAHoB+JqEiwEiAEYBDIphz5CpcpURVEy1nrqNmbDgE9eS1YNwixU3EA).
